src/NServiceBus.AcceptanceTests/Routing/When_registering_publishers_unobtrusive_messages.cs

```diff
@@ -0,0 +1,144 @@
+                    Assert.That(context.Subscribed, Is.True);
+                    Assert.That(context.ReceivedMessage, Is.True);
+                })
+                .Run();
+        }
+
+        public class Context : ScenarioContext
+        {
+            public bool Subscribed { get; set; }
+            public bool ReceivedMessage { get; set; }
+        }
+
+        public class Publisher : EndpointConfigurationBuilder
+        {
+            public Publisher()
+            {
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.OnEndpointSubscribed<Context>((e, ctx) => ctx.Subscribed = true);
+                    c.Conventions().DefiningEventsAs(t => t == typeof(SomeEvent));
+                });
+            }
+        }
+
+        public class SubscriberUsingRoutingApi : EndpointConfigurationBuilder
+        {
+            public SubscriberUsingRoutingApi()
+            {
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.Conventions().DefiningEventsAs(t => t == typeof(SomeEvent));
+
+                    var routing = new RoutingSettings<MessageDrivenPubSubTransportDefinition>(c.GetSettings());
+                    routing.RegisterPublisher(typeof(SomeEvent).Assembly, Conventions.EndpointNamingConvention(typeof(Publisher)));
+                }).IncludeType<SomeEvent>();
+            }
+
+            public class EventHandler : IHandleMessages<SomeEvent>
+            {
+                Context testContext;
+
+                public EventHandler(Context testContext)
+                {
+                    this.testContext = testContext;
+                }
+
+                public Task Handle(SomeEvent message, IMessageHandlerContext context)
+                {
+                    testContext.ReceivedMessage = true;
+                    return Task.FromResult(0);
+                }
+            }
+        }
+
+        public class SubscriberUsingEndpointMappings : EndpointConfigurationBuilder
+        {
+            public SubscriberUsingEndpointMappings()
+            {
+                EndpointSetup<DefaultServer>(c => c
+                .Conventions().DefiningEventsAs(t => t == typeof(SomeEvent)))
+                .AddMapping<SomeEvent>(typeof(Publisher))
+                .IncludeType<SomeEvent>();
+            }
+
+            public class EventHandler : IHandleMessages<SomeEvent>
+            {
+                Context testContext;
+
+                public EventHandler(Context testContext)
+                {
+                    this.testContext = testContext;
+                }
+
+                public Task Handle(SomeEvent message, IMessageHandlerContext context)
+                {
+                    testContext.ReceivedMessage = true;
+                    return Task.FromResult(0);
+                }
+            }
+        }
+
+        class MessageDrivenPubSubTransportDefinition : TransportDefinition, IMessageDrivenSubscriptionTransport
```

Comment:
to get to the parts of the routing API which are only accessible by transports implementing `IMessageDrivenSubscriptionTransport` (the publisher registration API) here: https://github.com/Particular/NServiceBus/pull/4030/files#diff-65aeb0d8e76e0a2cdbbd47c73fd6a283R77